File: custom_adapter/TopOnAdapter/Objc/6.5.x/Native/AlxTopOnNativeEvent.m
//
//  AlxTopOnNativeEvent.m
//  AlxAdsOCDemo
//

#import "AlxTopOnNativeEvent.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnBiddingRequestManager.h"

static NSString *const TAG = @"AlxTopOnNativeEvent";

@implementation AlxTopOnNativeEvent

- (void)nativeAdImpression:(AlxNativeAd *)nativeAd {
    NSLog(@"%@: nativeAdImpression", TAG);
    [self trackNativeAdImpression];
}

- (void)nativeAdClick:(AlxNativeAd *)nativeAd {
    NSLog(@"%@: nativeAdClick", TAG);
    [self trackNativeAdClick];
}

- (void)nativeAdClose:(AlxNativeAd *)nativeAd {
    NSLog(@"%@: nativeAdClose", TAG);
    [self trackNativeAdClosed];
}

- (NSDictionary *)createAssetWithNativeAd:(AlxNativeAd *)nativeAd {
    NSMutableDictionary *assetDic = [NSMutableDictionary dictionary];
    assetDic[kATAdAssetsCustomEventKey] = self;
    assetDic[kATAdAssetsCustomObjectKey] = nativeAd;
    
    // 原生广告自渲染：数据添加 / Native ad self-rendering: populate data
    assetDic[kATNativeADAssetsIsExpressAdKey] = @NO;
    assetDic[kATNativeADAssetsMainTitleKey] = nativeAd.title ?: @"";
    assetDic[kATNativeADAssetsMainTextKey] = nativeAd.desc ?: @"";
    assetDic[kATNativeADAssetsIconURLKey] = nativeAd.icon.url ?: @"";
    assetDic[kATNativeADAssetsImageURLKey] = nativeAd.images.firstObject.url ?: @"";
    assetDic[kATNativeADAssetsCTATextKey] = nativeAd.callToAction ?: @"";
    assetDic[kATNativeADAssetsAdvertiserKey] = nativeAd.adSource ?: @"";
    assetDic[kATNativeADAssetsLogoImageKey] = nativeAd.adLogo ?: @"";
    assetDic[kATNativeADAssetsContainsVideoFlag] = @(nativeAd.createType == ATNativeADSourceTypeVideo);
    
    if (nativeAd.images.count == 1) {
        AlxNativeAdImage *mainImage = nativeAd.images.firstObject;
        assetDic[kATNativeADAssetsMainImageWidthKey] = @(mainImage.width);
        assetDic[kATNativeADAssetsMainImageHeightKey] = @(mainImage.height);
    }
    
    return [assetDic copy];
}

- (NSString *)networkUnitId {
    return self.serverInfo[[AlxTopOnBaseManager unitID]] ?: @"";
}

- (NSMutableArray *)assets {
    return [NSMutableArray arrayWithObject:self.assetDict ?: @{}];
}

- (void)nativeAdFailToLoadWithDidFailWithError:(NSError * _Nonnull)error { 
    NSLog(@"%@: nativeAdFailToLoad", TAG);
    if (self.isC2SBiding) {
        [AlxTopOnBiddingRequestManager disposeLoadFailWithError:error
                                                         unitID:self.serverInfo[[AlxTopOnBaseManager unitID]]];
    } else {
        [self trackNativeAdLoadFailed:error];
    }
}

- (void)nativeAdLoadedWithDidReceive:(NSArray<AlxNativeAd *> * _Nonnull)ads { 
    NSLog(@"%@: nativeAdLoaded", TAG);
    
    AlxNativeAd *nativeAd = ads.firstObject;
    if (!nativeAd) {
        NSString *errorStr = @"native ad data is empty";
        NSLog(@"%@: native ad data is empty", TAG);
        NSError *error = [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
        
        if (self.isC2SBiding) {
            [AlxTopOnBiddingRequestManager disposeLoadFailWithError:error
                                                             unitID:self.serverInfo[[AlxTopOnBaseManager unitID]]];
        } else {
            [self trackNativeAdLoadFailed:error];
        }
        return;
    }
    
    NSDictionary *assetData = [self createAssetWithNativeAd:nativeAd];
    
    if (self.isC2SBiding) {
        self.assetDict = assetData;
        [AlxTopOnBiddingRequestManager disposeLoadSuccessWithPrice:[nativeAd getPrice]
                                                            unitID:self.serverInfo[[AlxTopOnBaseManager unitID]]];
        self.isC2SBiding = NO;
    } else {
        [self trackNativeAdLoaded:@[assetData]];
    }
}

@end

File: custom_adapter/TopOnAdapter/Swift/6.4.x/AlxTopOnNativeEvent.swift
//
//  AlxTopOnNativeEvent.swift
//

import Foundation
import AnyThinkSDK
import AnyThinkNative
import AlxAds

@objc(AlxTopOnNativeEvent)
public class AlxTopOnNativeEvent: ATNativeADCustomEvent,AlxNativeAdLoaderDelegate,AlxNativeAdDelegate {
    
    private static let TAG = "AlxTopOnNativeEvent"
    
    public var assetDict:[AnyHashable:Any] = [:]
    
    public override init(info serverInfo: [AnyHashable : Any], localInfo: [AnyHashable : Any]) {
        super.init(info: serverInfo, localInfo: localInfo)
    }
    
    public func nativeAdLoaded(didReceive ads: [AlxNativeAd]) {
        NSLog("%@: nativeAdLoaded",AlxTopOnNativeEvent.TAG)
        
        guard let nativeAd = ads.first else{
            let errorStr = "native ad data is empty"
            NSLog("%@: native ad data is empty",AlxTopOnNativeEvent.TAG)
            let error = AlxTopOnBaseManager.error(code: -100, msg: errorStr)
            if self.isC2SBiding {
                AlxTopOnBiddingRequestManager.disposeLoadFail(error: error, unitID: self.serverInfo[AlxTopOnBaseManager.unitID] as? String)
            }else{
                self.trackNativeAdLoadFailed(error)
            }
            return
        }
        
        let assetData = self.createAsset(nativeAd: nativeAd)
        if self.isC2SBiding {
            self.assetDict = assetData
            AlxTopOnBiddingRequestManager.disposeLoadSuccess(price: nativeAd.getPrice(), unitID: self.serverInfo[AlxTopOnBaseManager.unitID] as? String)
            self.isC2SBiding = false
        }else{
            self.trackNativeAdLoaded([assetData])
        }
    }
    
    public func nativeAdFailToLoad(didFailWithError error: Error) {
        NSLog("%@: nativeAdFailToLoad",AlxTopOnNativeEvent.TAG)
        if self.isC2SBiding {
            AlxTopOnBiddingRequestManager.disposeLoadFail(error: error, unitID: self.serverInfo[AlxTopOnBaseManager.unitID] as? String)
        }else{
            self.trackNativeAdLoadFailed(error)
        }
    }
    
    public func nativeAdImpression(_ nativeAd:AlxNativeAd){
        NSLog("%@: nativeAdImpression",AlxTopOnNativeEvent.TAG)
        self.trackNativeAdImpression()
    }
    
    public func nativeAdClick(_ nativeAd:AlxNativeAd){
        NSLog("%@: nativeAdClick",AlxTopOnNativeEvent.TAG)
        self.trackNativeAdClick()
    }
    
    public func nativeAdClose(_ nativeAd:AlxNativeAd){
        NSLog("%@: nativeAdClose",AlxTopOnNativeEvent.TAG)
        self.trackNativeAdClosed()
    }
    
    public func createAsset(nativeAd:AlxNativeAd)->[AnyHashable:Any]{
        var assetDic:[AnyHashable:Any] = [:]
        assetDic[kATAdAssetsCustomEventKey] = self
        assetDic[kATAdAssetsCustomObjectKey] = nativeAd
        
        //原生广告自渲染：数据添加
        assetDic[kATNativeADAssetsIsExpressAdKey] = false
        assetDic[kATNativeADAssetsMainTitleKey] = nativeAd.title
        assetDic[kATNativeADAssetsMainTextKey] = nativeAd.desc
        assetDic[kATNativeADAssetsIconURLKey] = nativeAd.icon?.url
        assetDic[kATNativeADAssetsImageURLKey] = nativeAd.images?.first?.url
        assetDic[kATNativeADAssetsCTATextKey] = nativeAd.callToAction
        assetDic[kATNativeADAssetsAdvertiserKey] = nativeAd.adSource
        assetDic[kATNativeADAssetsLogoImageKey] = nativeAd.adLogo
        assetDic[kATNativeADAssetsContainsVideoFlag] = nativeAd.createType == AlxNativeAd.Create_Type_Video ? true : false
        
        if nativeAd.images?.count == 1,let mainImage = nativeAd.images?.first {
            assetDic[kATNativeADAssetsMainImageWidthKey] = mainImage.width
            assetDic[kATNativeADAssetsMainImageHeightKey] = mainImage.height
        }
        
        return assetDic
    }
    
    public override var networkUnitId: String {
        get {
            return self.serverInfo[AlxTopOnBaseManager.unitID] as? String ?? ""
        }
        set {
        }
    }
    
    public override var assets: NSMutableArray{
        get {
            return [self.assetDict]
        }
        set {
        }
    }
    
}
